### PR TITLE
Add `auto_capture` option to AnimationPlayer

### PR DESCRIPTION
**Pull request by TokageItLab**
Add `auto_capture` option to AnimationPlayer.

image

I have received a request from @YeldhamDev to perform capture by `play()`.

My concern is that `play()` is used by many internal APIs, so it is risky to hardcode capture to work there.

Also, the AnimationNode that performs some of the playback of the AnimationTree does not automatically handle capture, so if AnimationPlayer did it, it would have to be explicit.

This PR adds an `auto_capture` option to Playback Option.

I found out during the implementation that it is quite similar to `default_blend_time`, so it makes sense to have it in close proximity. It seems a bit funny that double interpolation is performed when both are enabled, but computationally it seems to work correctly.

Before, I thought about adding them to the `play()` argument, but didn't do it. Because making them properties of the AnimationPlayer makes them explicit and allows the user to control the animation as it is being executed. Also, there are no compatibility issues.

For example, when queueing playback of a capture animation, there may be cases where the user wants capture to occur at all times during the animation, or only at the first playback. The latter can be implemented by the user disabling the `auto_capture` option after the first playback.

This PR revives compatibility with the old capture behavior without replacing `play()` with `play_with_capture()`. (Note that there is no compatibility for cubic interpolation.)

Then, `play_with_capture()` is left for more detailed options; For `auto_capture`, options like `default_capture_duration`, `default_capture_tween_transition_type` could be added in the same way as `dafault_blend_time`, but it is possible to follow up later on demands.

[capture_demo.zip](https://github.com/godotengine/godot/files/15183264/capture_demo.zip)

### CODE REVIEW

**Review by YeldhamDev — APPROVED** [review]
Seems to be working well, great job!

On a slightly unrelated note, the docs on `play_with_capture()` states:
> If name is blank, it specifies [member assigned_animation].

Besides "name" needing to be changed to "[param name]", `name` should likely allow to be left blank, like its `play()` counterpart.

**Comment by TokageItLab** [comment]
> Besides "name" needing to be changed to "[param name]", name should likely allow to be left blank, like its play() counterpart.

Fixed, thank you for testing!

File: doc/classes/Animation.xml [hunk 1]
```diff
@@ -595,6 +595,9 @@
 		</method>
 	</methods>
 	<members>
+		<member name="capture_included" type="bool" setter="_set_capture_included" getter="is_capture_included" default="false">
+			Returns [code]true[/code] if the capture track is included. This is cached readonly value for performance.
```
Comment (AThousandShips) [inline doc/classes/Animation.xml:599]:
```suggestion
			Returns [code]true[/code] if the capture track is included. This is a cached readonly value for performance.
```

File: doc/classes/Animation.xml [hunk 2]
```diff
@@ -658,7 +661,7 @@
 			Update at the keyframes.
 		</constant>
 		<constant name="UPDATE_CAPTURE" value="2" enum="UpdateMode">
-			Same as [constant UPDATE_CONTINUOUS] but works as a flag to capture the value of the current object and perform interpolation in some methods. See also [method AnimationMixer.capture] and [method AnimationPlayer.play_with_capture].
+			Same as [constant UPDATE_CONTINUOUS] but works as a flag to capture the value of the current object and perform interpolation in some methods. See also [method AnimationMixer.capture], [member AnimationPlayer.playback_auto_capture] and [method AnimationPlayer.play_with_capture].
```
Comment (AThousandShips) [inline doc/classes/Animation.xml:664]:
```suggestion
			Same as [constant UPDATE_CONTINUOUS] but works as a flag to capture the value of the current object and perform interpolation in some methods. See also [method AnimationMixer.capture], [member AnimationPlayer.playback_auto_capture], and [method AnimationPlayer.play_with_capture].
```

File: scene/animation/animation_player.cpp [hunk 3]
```diff
@@ -525,6 +473,47 @@ void AnimationPlayer::play(const StringName &p_name, double p_custom_blend, floa
...
 }
 
+void AnimationPlayer::_capture(const StringName &p_name, bool p_from_end, double p_duration, Tween::TransitionType p_trans_type, Tween::EaseType p_ease_type) {
+	StringName name = p_name;
+	if (name == StringName()) {
+		name = playback.assigned;
+	}
+
+	Ref<Animation> anim = get_animation(name);
+	if (!anim.is_valid() || !anim->is_capture_included()) {
```
Comment (AThousandShips) [inline scene/animation/animation_player.cpp:483]:
```suggestion
	if (anim.is_null() || !anim->is_capture_included()) {
```

File: scene/resources/animation.cpp [hunk 4]
```diff
@@ -966,6 +967,28 @@ void Animation::remove_track(int p_track) {
...
+	capture_included = false;
+	for (int i = 0; i < tracks.size(); i++) {
+		if (tracks[i]->type == TYPE_VALUE) {
+			ValueTrack *vt = static_cast<ValueTrack *>(tracks[i]);
+			if (vt->update_mode == UPDATE_CAPTURE) {
+				capture_included = true;
+				break;
+			}
+		}
+	};
```
Comment (AThousandShips) [inline scene/resources/animation.cpp:991]:
```suggestion
	}
```

File: scene/animation/animation_player.cpp [hunk 5]
```diff
@@ -815,9 +812,12 @@ void AnimationPlayer::_bind_methods() {
...
 
+	ClassDB::bind_method(D_METHOD("set_auto_capture", "auto_capture"), &AnimationPlayer::set_auto_capture);
+	ClassDB::bind_method(D_METHOD("is_auto_capture"), &AnimationPlayer::is_auto_capture);
+
 	ClassDB::bind_method(D_METHOD("play", "name", "custom_blend", "custom_speed", "from_end"), &AnimationPlayer::play, DEFVAL(StringName()), DEFVAL(-1), DEFVAL(1.0), DEFVAL(false));
 	ClassDB::bind_method(D_METHOD("play_backwards", "name", "custom_blend"), &AnimationPlayer::play_backwards, DEFVAL(StringName()), DEFVAL(-1));
-	ClassDB::bind_method(D_METHOD("play_with_capture", "name", "duration", "custom_blend", "custom_speed", "from_end", "trans_type", "ease_type"), &AnimationPlayer::play_with_capture, DEFVAL(-1.0), DEFVAL(-1), DEFVAL(1.0), DEFVAL(false), DEFVAL(Tween::TRANS_LINEAR), DEFVAL(Tween::EASE_IN));
+	ClassDB::bind_method(D_METHOD("play_with_capture", "name", "duration", "custom_blend", "custom_speed", "from_end", "trans_type", "ease_type"), &AnimationPlayer::play_with_capture, DEFVAL(StringName()), DEFVAL(-1.0), DEFVAL(-1), DEFVAL(1.0), DEFVAL(false), DEFVAL(Tween::TRANS_LINEAR), DEFVAL(Tween::EASE_IN));
```
Comment (AThousandShips) [inline scene/animation/animation_player.cpp:820]: I'm unsure why this doesn't trigger the API check, that might be a bug

**Review by AThousandShips — APPROVED** [review]

File: scene/animation/animation_player.cpp (hunk 5 above)
Comment (TokageItLab) [reply scene/animation/animation_player.cpp:820]: If you are implying compatibility, it is because this API did not yet exist in 4.2.

Comment (AThousandShips) [reply scene/animation/animation_player.cpp:820]: Oh then it's fine, thank you

**Comment by akien-mga** [comment]
Thanks!

**Comment by KoBeWi** [comment]
`capture_included` will stay in the animation after track is changed back to other update mode. Adding or removing a track fixes this.

File: scene/resources/animation.cpp [hunk 6]
```diff
@@ -3870,9 +3895,13 @@ void Animation::_bind_methods() {
 
 	ClassDB::bind_method(D_METHOD("compress", "page_size", "fps", "split_tolerance"), &Animation::compress, DEFVAL(8192), DEFVAL(120), DEFVAL(4.0));
 
+	ClassDB::bind_method(D_METHOD("_set_capture_included", "capture_included"), &Animation::set_capture_included);
+	ClassDB::bind_method(D_METHOD("is_capture_included"), &Animation::is_capture_included);
+
 	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "length", PROPERTY_HINT_RANGE, "0.001,99999,0.001,suffix:s"), "set_length", "get_length");
 	ADD_PROPERTY(PropertyInfo(Variant::INT, "loop_mode", PROPERTY_HINT_ENUM, "None,Linear,Ping-Pong"), "set_loop_mode", "get_loop_mode");
 	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "step", PROPERTY_HINT_RANGE, "0,4096,0.001,suffix:s"), "set_step", "get_step");
+	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "capture_included", PROPERTY_HINT_NONE, "", PROPERTY_USAGE_STORAGE | PROPERTY_USAGE_READ_ONLY | PROPERTY_USAGE_NO_EDITOR), "_set_capture_included", "is_capture_included");
```
Comment (KoBeWi) [inline scene/resources/animation.cpp:3904]: `PROPERTY_USAGE_NO_EDITOR` and `PROPERTY_USAGE_STORAGE` have the same value. You should include only one of them.

File: doc/classes/AnimationPlayer.xml [hunk 7]
```diff
@@ -210,6 +211,9 @@
 			If [code]true[/code] and the engine is running in Movie Maker mode (see [MovieWriter]), exits the engine with [method SceneTree.quit] as soon as an animation is done playing in this [AnimationPlayer]. A message is printed when the engine quits for this reason.
 			[b]Note:[/b] This obeys the same logic as the [signal AnimationMixer.animation_finished] signal, so it will not quit the engine if the animation is set to be looping.
 		</member>
+		<member name="playback_auto_capture" type="bool" setter="set_auto_capture" getter="is_auto_capture" default="true">
+			If [code]true[/code], performs [method AnimationMixer.capture] before playback automatically. This means just [method play_with_capture] is executed with default arguments instead of [method play].
```
Comment (KoBeWi) [inline doc/classes/AnimationPlayer.xml:215]: I think this should clarify that it only happens when the animation has a capture track.

**Pull request merged**